tritium-lib/src/main/java/com/palantir/tritium/proxy/ByteBuddyInstrumentation.java

```diff
@@ -0,0 +1,234 @@
+                    builder = builder.defineMethod(delegateMethod(iface), iface, Modifier.PRIVATE | Modifier.FINAL)
+                            .intercept(FieldAccessor.ofField("delegate")
+                                    .withAssigner(Assigner.DEFAULT, Assigner.Typing.DYNAMIC));
+                }
+                for (Method method : iface.getMethods()) {
+                    int index = allMethods.size();
+                    allMethods.add(method);
+                    // Retain tritium proxy detail where hashcode, equals, and toString cannot be instrumented.
+                    builder = builder.method(ElementMatchers.not(ElementMatchers.isHashCode()
+                            .or(ElementMatchers.isEquals())
+                            .or(ElementMatchers.isToString()))
+                            .and(ElementMatchers.is(method)))
+                            .intercept(Advice.withCustomMapping()
+                                    .bind(ByteBuddyInstrumentationAdvice.MethodIndex.class, index)
+                                    .to(ByteBuddyInstrumentationAdvice.class)
+                                    .wrap(allowDirectAccess
+                                            ? MethodDelegation.toField("delegate")
+                                            : MethodDelegation.toMethodReturnOf(delegateMethod(iface))));
+                }
+            }
+            return builder
+                    .defineField("delegate", interfaceClass,
+                            Opcodes.ACC_PRIVATE | Opcodes.ACC_FINAL)
+                    .defineField("invocationEventHandler", InvocationEventHandler.class,
+                            Opcodes.ACC_PRIVATE | Opcodes.ACC_FINAL)
+                    .defineField("instrumentationFilter", InstrumentationFilter.class,
+                            Opcodes.ACC_PRIVATE | Opcodes.ACC_FINAL)
+                    .defineField("methods", Method[].class, Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC)
+                    .initializer(new StaticFieldLoadedTypeInitializer("methods", allMethods.toArray(new Method[0])))
+                    .make()
+                    .load(classLoader, ClassLoadingStrategy.Default.WRAPPER)
+                    .getLoaded();
+        });
+    }
+
+    private static ClassLoader getClassLoader(Class<?> clazz) {
+        ClassLoader loader = clazz.getClassLoader();
+        // Some internal classes return a null classloader, in these cases we provide the instrumentation loader.
+        if (loader == null) {
+            return ByteBuddyInstrumentation.class.getClassLoader();
+        }
+        return loader;
+    }
+
+    private static boolean isAccessible(Class<?> clazz) {
+        // Base case for enclosing class accessibility check
+        if (clazz == null) {
+            return true;
+        }
+        return (clazz.isInterface() || Modifier.isPublic(clazz.getModifiers()))
+                && isAccessible(clazz.getEnclosingClass());
```

Comment:
What happens for interfaces defined in a non-interface class?

Reply:
This recursively checks that enclosing classes are public, otherwise we cannot implement the interface.

bug: I also need to check if interfaces are public, InstrumentationTest happens to pass because it's in the same package as `ByteBuddyInstrumentation`.

Reply:
Adding a slew of tests for these scenarios.